### PR TITLE
#166840920 User Authentication with OAuth2.0

### PR DESCRIPTION
**Pull request by tolumide-ng**
#### What does this PR do?
Enables user signup with google, facebook and gtihub

#### Description of Task to be completed?
- Make a request to **localhost/<port>/auth/github** and approve to signup with Github
- Make a request to **localhost/auth/<port>/facebook** and approve to signup with Facebook
- Make a request to **localhost/auth/<port>/google** and approve to signup with google

#### How should this be manually tested?
Clone the repo, cd into it and run **npm run start:dev**, makes a request to **localhost:<PORT>/auth/google**, and choose the email address you want to signup with

#### Any background context you want to provide?
Provide all details required as specified in .env.example

#### What are the relevant pivotal tracker stories?
#166840920

### CODE REVIEW

**Review by Onyimatics — APPROVED** [review]
Nice work @tolumide-ng. Looks good to me.

File: src/helpers/passport/callback.js [hunk 1]
```diff
@@ -0,0 +1,26 @@
+import db from '../../database/models';
+
+const callback = async (accessToken, refreshToken, payload, done) => {
+  const { displayName, emails, photos } = payload;
+  const email = emails[0].value;
+  try {
+    const confirmUser = await db.User.findOrCreate({
```
Comment (segunolalive) [inline src/helpers/passport/callback.js]: Please use the BaseRepository instead of calling the user model directly

Comment (segunolalive) [inline src/helpers/passport/callback.js]: Please use the BaseRepository.

File: src/helpers/passport/facebook.js [hunk 2]
```diff
@@ -0,0 +1,28 @@
...
+import callback from './callback';
+
+dotenv.config();
+
+passport.serializeUser((user, done) => {
+  done(null, user.id);
+});
+
+passport.deserializeUser((id, done) => {
+  db.User.findById(id).then(user => done(null, user));
```
Comment (segunolalive) [inline src/helpers/passport/facebook.js]: Avoid calling the models directly

File: src/index.js [hunk 3]
```diff
@@ -4,10 +4,27 @@ import cors from 'cors';
...
+import Routes from './routes/v1/index';
+import './helpers/passport/google';
+import './helpers/passport/github';
 
 dotenv.config();
 
 const app = express();
+
+app.use(
+  cookieSession({
```
Comment (segunolalive) [inline src/index.js]: why are we using cookies/sessions

File: src/test/oauth.spec.js [hunk 4]
```diff
@@ -0,0 +1,73 @@
...
+
+describe('SOCIAL LOGIN', () => {
+  it('should hit the facebook URL for social login', async () => {
+    nock('https://www.facebook.com/')
+      .filteringPath(() => '/auth/facebook')
+      .get('/auth/facebook')
+      .reply(200, { message: 'hit the route' });
+
+    const res = await chai.request(app).get('/auth/facebook');
+    expect(res.body.message).to.include('hit the route');
```
Comment (segunolalive) [inline src/test/oauth.spec.js:15]: why are we sending this message to users? 🤷‍♂ 

I don't think it provides any business value

File: src/test/oauth.spec.js [hunk 5]
```diff
@@ -0,0 +1,73 @@
...
+  });
+
+  it('should call the github call back', async () => {
+    nock('https://www.github.com/')
+      // .filteringPath(() => '/auth/github/callback')
+      .get('/auth/github/callback')
+      .reply(200, { message: 'hit the route' });
+
+    const res = await chai.request(app).get('/auth/github/callback');
+    expect(res.body).to.an('object');
```
Comment (segunolalive) [inline src/test/oauth.spec.js:68]: `res.body` is always an object. That's what `bodyParser` does. This doesn't test that your code functions as it should. Remove it and test what that your route returns the appropriate data

**Review by segunolalive — COMMENTED** [review]
👍  @tolumide-ng 

Check out the feedback

File: .coveralls.yml [hunk 6]
```diff
@@ -1 +1,3 @@
 repo_token: COVERALLS_REPO_TOKEN
+exclude_patterns:
```
Comment (timi-codes) [inline .coveralls.yml:2]: Thanks for adding this.

File: src/controllers/user.controller.js [hunk 7]
```diff
@@ -145,7 +142,7 @@ class UserController {
    * @memberof UserController
    */
   static async verifyUser(req, res) {
-    const { email } = req.token;
+    const { email } = req.currentUser;
```
Comment (timi-codes) [inline src/controllers/user.controller.js:145]: Thanks for renaming

File: src/database/config/config.js [hunk 8]
```diff
@@ -1,12 +1,12 @@
-const config = require('dotenv');
```
Comment (timi-codes) [inline src/database/config/config.js:1]: Is there any reason why you chose to rename this?

File: src/database/config/config.js [hunk 9]
```diff
@@ -19,10 +19,10 @@ module.exports = {
...
   },
   test: {
-    username: process.env.DB_USERNAME,
-    password: process.env.DB_PASSWORD,
-    database: process.env.DB_DATABASE,
-    host: process.env.DB_HOST,
+    username: process.env.DB_TEST_USERNAME,
+    password: process.env.DB_TEST_PASSWORD || '',
+    database: process.env.DB_TEST_DATABASE,
+    host: process.env.DB_TEST_HOST || '127.0.0.1',
```
[the remark this answers is not in the document]
Comment (timi-codes) [reply src/database/config/config.js]: I think there is no reason doing '|| 127.0.0.1' localhost will always be used to run your test.

**Review by timi-codes — CHANGES_REQUESTED** [review]
This appears good to me. However, I left some comments

File: src/database/config/config.js (hunk 8 above)
Comment (tolumide-ng) [reply src/database/config/config.js:1]: Yes @timi-codes , I did this for consistency. There are other parts in our codebase that import `dotenv` and use it this way

**Pull request merged**